File: cpp/100. Same Tree.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
// Recursion, Time: O(n), Space: O(n)
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if(!p && !q) return true;
        if((!p && q) || (p && !q)) return false;
        if(p->val != q->val)  return false;
        return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);      
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    
}

File: python/100-Same Tree.py
from bintree_helper import *

class Solution:
    def isSameTree(self, p, q) -> bool:
        "DFS, Time: O(n), Space: O(n)"
        def dfs(l, r):
            if l is None and r is None: return True
            if l is None or r is None: return False
            return l.val == r.val and dfs(l.left, r.left) and dfs(l.right, r.right)
        return dfs(p, q)
    def isSameTree(self, p, q) -> bool:
        "BFS, Time: O(n), Space: O(n)"
        from collections import deque
        q1 = deque([p])
        q2 = deque([q])
        while q1 and q2:
            l = q1.popleft()
            r = q2.popleft()
            if l is None and r is None: continue
            if l is None or r is None: return False
            if l.val != r.val: return False
            q1.append(l.left)
            q1.append(l.right)
            q2.append(r.left)
            q2.append(r.right)
        if len(q1) == 0 and len(q2) == 0: return True
        return False

p = deserialize("[1,2,3]")
q = deserialize("[1,2,3]")
result = Solution().isSameTree(p, q)
assert result == True

p = deserialize("[1,2]")
q = deserialize("[1,null,2]")
result = Solution().isSameTree(p, q)
assert result == False

p = deserialize("[1,2,1]")
q = deserialize("[1,1,2]")
result = Solution().isSameTree(p, q)
assert result == False